### DISCUSSION TITLE
SSsq/SShd

### DISCUSSION THREAD

**Discussion by Illeprih** [vgmtrans/vgmtrans#478] [Music formats]
Hello, would you be interested in adding support for the PS1 SSsq/SShd format? I have done fairly extensive research into it and was able to achieve playback. Here are my notes on the file strucutres:

SSsq
000h	2				Master Volume
002h	2				Ticks per quarter note
004h	2				Tempo
006h	6				Zero filled
00Ch	4				Magic "SSsq"
010h	10h*10h			10h Channels
110h	...				Sequence (This format seems to allow only for one. I've only seen more in a format where the SSsq is embeded inside SShd)

Channels
000h	1				Unknown - always 0?
001h	1				Index/ID
002h	1				Program index (from SShd)
003h	1				SSsq volume
004h	1				Pan
005h 	4				Unknown (Probably ADSR settings)
009h 	1				Modulation
00Ah 	1				Pitch Bend
00Bh 	1				Priority
00Ch 	1				Breath
00Dh 	1				Unknown (Maybe sixteenths?)
00Eh 	1				Adjusted volume (When channel is issued change volume command, it applies both the value and the SSsq value, however LoD seem to override SSsq value on load)
00Fh 	1				Unknown

SShd
000h 	4				SShd size
004h 	4				Soundbank size
008h 	4				Unknown - zerofilled
00Ch 	4				Magic "SShd"
010h 	1Ch*4			Ptr table (SShd subfiles)
080h 	...				Data

SShd subfiles (As they are in LoD, but I suspect it should be the same for other games aswell)

0 - Programs (For regular music, otherwise it's part of the embeded SSsq)
1 - Velocity Table
2 - Breath Controls
3 - Embeded Sequences (for SFX formats)
4 - Embeded SSsq (for SFX formats)

Programs (Subfile 0)
000h	2				Instrument upper bound (count - 1)
002h	Count*2			Instrument relative offset (FFFFh if unused, seems kinda stupid, but it does happen)
...		...				Upper bound + 1 Instrument

Instrument
000h	1				Type + Tones upper bound (count - 1). (FFh are SFX instruments, 8Xh are instruments, that can play multiple tones on a single command, otherwise 0, second nibble is upper bound of notes AKA count - 1)
001h	1				Volume
002h	1				Pan
003h	1				Unknown/Unused
004h	1				Pitch bend multiplier
005h	1				Breath control index (if there are multiple breath control tables in SShd subfile 2)
006h	1				Starting note (only used for SFX instruments)
007h	1				SFX Tone upper bound
008h	Tones*10h		Tone data

Tone
000h	1				noteMin
001h	1				noteMax (If a KeyOn is called on a specific program, notes are selected if they are in range)
002h	1				keyRoot (the offset from the note requested is used to pitch bend)
003h	1				sixteenths (flat pitch bend in 1/16 of a semitone) [This can be somewhat easily converted to 64ths for higher playback precision, I guess they felt constrained by the size of lookup tables, however the PSX tables are awfully inefficient]
004h	2				sound bank ptr (*4)
006h	4				ADSR
00Ah	1				Volume override (if used, this value would be << 8 and volume would be | onto the lower bits during volume calculation)
00Bh	1				Volume
00Ch	1				Pan
00Dh	1				Pitch bend multiplier
00Eh	1				Breath control Index
00Fh	1				Flags (1 - high priority, 2 - noise, 10h, pitch bend multi from instrument, 20h - modulation on, 40h - breath control index from instrument, 80h - reverb on

Velocity Table (Subfile 1)
000h	2				Unused, maybe at some point, they wanted to support multiple velocity tables?
002h	80h				Volumes for velocity 0-7Fh

Breath controls (Subfile 2)
000h	2				Breath controls upper bound (count - 1)
002h	Count*2			Relative offsets
...		...				Breath values

Embeded Sequences (Subfile 3)
000h	2				Sequece pack upper bound (count - 1)
002h	PackCount*2		Sequece pack relative offsets
[						
...		2				Sequence upper bound (count - 1)
...		SeqCount*2		Sequence relative offsets
] * pack count
...		...				Sequence Data

Embeded SSsq (Subfile 4)
000h	2				Master volume
002h	Eh				Rest of the SSsq header, but zerofilled
010h	18h*10h			18h Channels (I currently have no idea how are the ones after 10h supposed to be addressed, when the index is a single nibble)
400h	...				Single Instrument with n (Instrument byte 007h) SFX Tones

**Comment by jayare5** [comment]
Is this the same format in this page?
https://problemkaputt.de/psxspx-cdrom-file-audio-other-formats.htm

I would LOVE to have support for Wild Arms 2 and Alundra 2!!

**Reply by Illeprih** [reply]
Yes, it's the .HD/.SQ format.
One small issue might be that for sequences, it uses two different specs with different sequence cmd lengths.

**Comment by MarkusMaal** [comment]
I've been messing around with the SShd format specifically, it seems like the ADSR envelope for each instrument's tone (a.k.a. split chunk) is stored exactly the same way as described in the "SPU2 Overview" PDF document (specifically Chapter 3.2: "ADSR1 / ADSR2 : Envelope" and Chapter 4.1: "Rate Parameter Table").

It's divided into two 16-bit bitfields called ADSR1 and ADSR2. These control the following parameters:

**ADSR1** (u16 LE)
* Sustain Level ((n+1)/16) : 4
* Decay Rate (index of ms) : 4 (always -exponential)
* Attack Rate (index of ms) : 7
* Mode for Attack Rate (+linear, +exponential) : 1

**ADSR2** (u16 LE)
* Release Rate (index of ms) : 5
* Mode for Release Rate (-linear, -exponential) : 1
* Sustain Rate (index of ms) : 7 (don't think there is a direct equivalent in SF2 or DLS specifications)
* Mode for Sustain Rate (+linear, -linear, +exponential, -exponential) : 2
* Empty/Reserved : 1

If sustain rate is 7Fh, it should be ignored. Similarly, if release rate is 1Fh, assume it's 0.0.

I have a basic implementation of the [ADSR decoding algorithm](https://github.com/MarkusMaal/FlipnicFileTools/blob/b360be047436e057ef16c1c0f7433e50bbedf4d7/FlipnicLib/Formats/Jam/JamProgChunk.cs#L169-L306) and [the actual SF2 generators](https://github.com/MarkusMaal/FlipnicFileTools/blob/b360be047436e057ef16c1c0f7433e50bbedf4d7/FlipnicLib/Formats/Jam/Converter.cs#L210-L217) in my FlipnicFileTools repository (it takes Standard MIDI files instead of SSsq and is written in C#.NET). It's not perfect, since I'm using brute-forced values in certain places (also it maybe possible I typed the constant values wrong, since I haven't thoroughly checked them), but it should be a good starting point.

**Reply by Illeprih** [reply]
The ADSR is gonna be constrained to essentially a single spec. The sequencer can be provided by the SW, but ADSR is tied pretty closely to the HW itself.

I've personally stayed away from trying to make a convertor to SF2, because I couldn't figure out a way to implement few of the features, like vibratos, which are directly linked with the instruments.

**Reply by mikelow** [reply]
@MarkusMaal in case you weren't aware, VGMTrans [does handle](https://github.com/vgmtrans/vgmtrans/blob/e9ff182ff74a21ed81d4062bd90f0692860c0d64/src/main/formats/common/PSXSPU.h#L127) PS1 and PS2 ADSR envelopes, although it is some of the earliest code and is very sloppy.

If you're aware of flaws in the logic by all means let us know or put in a PR.

@Illeprih The only major deficiency in the SF2 ADSR spec is a lack of "sustain rate", sometimes called a second decay rate. Both the SNES and PS1/PS2 SPUs support this curve (no coincidence - they were designed by the same person). Another difference: PS1/PS2 decay/release rates can be made either linear or exponential. This is less consequential as we do our best to compensate, but still significant.

SF2 supports baking vibrato and tremelo into instrument data. It's limited to a frequency, depth, and delay. This was first used recently in VGMTrans for the Sega Saturn format. See [here](https://github.com/vgmtrans/vgmtrans/blob/e9ff182ff74a21ed81d4062bd90f0692860c0d64/src/main/formats/SegSat/SegSatInstrSet.cpp#L394) for the format usage and [here](https://github.com/vgmtrans/vgmtrans/blob/e9ff182ff74a21ed81d4062bd90f0692860c0d64/src/main/conversion/SF2File.cpp#L377) for the SF2 file implementation.

**Reply by Illeprih** [reply]
In this format, it is defined by a little embedded file that defines the specific profile. You get 60 values per loop, but you can make your own. From the game files I've seen, it tends to use standardized ones, but it still means you can't just generalize it easily. The most used one is just a sine wave, but some files actually contain a bug, where it was incorrectly calculated as 59 values so the first and last value are the same. The code implementation of looping over this file was also very cursed and bug ridden and would not loop correctly with some frequency values.

**Comment by mikelow** [comment]
@Illeprih thanks for the info on the format! I'm very glad to have it documented here. I can't promise I'll implement it due to lack of time and priorities, but you are, of course, very welcome to contribute a PR for it.

**Reply by Illeprih** [reply]
After all, it's a very niche format and only used by a couple of games. My hunch is that it was more of a prototype format made to inform their decisions on what was to be the format for PS2.

I've toyed with the idea of writing a PR, but am yet to find the time and will to deal with C++. Do feel free to ask, if something's missing, in the eventuality that you get to it before I do.
